### DISCUSSION TITLE
QOI - The “Quite OK Image Format” for fast, lossless image compression

### DISCUSSION THREAD

**Discussion by wilberforce** [Moddable-OpenSource/moddable#764] [Ideas]
@phoddie 

This looks like a great fit for Moddable.
https://phoboslab.org/log/2021/11/qoi-fast-lossless-image-compression

https://qoiformat.org/
https://github.com/phoboslab/qoi

QOI is fast. It losslessy compresses images to a similar size of PNG, while offering 20x-50x faster encoding and 3x-4x faster decoding.

QOI is simple. The reference en-/decoder fits in about 300 lines of C. The file format specification is a single page PDF.

The O(n) decoding would suit the micros and might speed up animations.

Someone has done (an out of date) typescript port:
https://github.com/panzi/jsqoi

**Comment by phoddie** [comment]
Interesting find. Some notes:

- It looks like it could offer a nice balance of feature for some situations.
- It should be possible to integrate with Poco rendering, though it would mostly be a rewrite of the decoder since the reference implementation isn't a streaming decoder.
- Clipping requires parsing through all pixels to the top and left of the clip origin, so not optimal for an atlas of UI elements.
- It uses 8-bit per channel to store colors. That will work, but most of our targets are 565, so each fully specified color uses one more byte of data than will ever be displayed. A variation of the image format could optimize that.
- It won't be faster than what's running on device now for UI rendering (we would never try to render a PNG directly -- both because of speed and memory - except in exceptional situations [where PNG compatibility is the priority](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/commodetto/pngdisplay/main.js).

**Comment by phoddie** [comment]
Thinking about this some more... the compressed bit stream is really tuned for pixels that have 8-bits per component. For example, the big delta packet `QOI_OP_LUMA` is 16-bits, 14 of which are the payload. For 16-bit pixels, that is just 2 bits short of an entire pixel. The small delta packet `QOI_OP_DIFF` adjusts the low two bits of each color component. When displaying an 8-bit per component image on a 565 display, those bits will never be visible. And, of course, the color packet `QOI_OP_RGB` is 8-bits per component, which is one byte bigger than needed.

So, if used as-is, the main compression benefits are from constant color runs and using `QOI_OP_INDEX` to reduce recently used pixels to 8 bits. The benefit depends on the image, but it seems likely to deliver around 50% compression (better for images with solid color runs).

A 16-bit 565 variant (or 555) could do better:

- `QOI_OP_RGB` would save one byte
- `QOI_OP_DIFF`could be re-biased to get the low bits of the components
- `QOI_OP_LUMA` might still be useful (once re-biased) but feels less so.

Do you know if someone has already investigated applying this to 16-bit pixels?

**Reply by wilberforce** [reply]
> Do you know if someone has already investigated applying this to 16-bit pixels?

Sorry no - I just saw this as a news item.

There is also this Experimental branch: https://github.com/phoboslab/qoi/issues/71

We could raise an issue as a starting point for discussing a 16 bit version.

**Comment by phoddie** [comment]
Asking about it in their repository makes more sense than just starting to hack at the problem. ;) Would you mind doing that?

**Comment by wilberforce** [comment]
@phoddie 
New kid on the block - supports 565.

https://github.com/bitbank2/SLIC

**Reply by phoddie** [reply]
I had a chance to look at SLIC. It is a better fit for what we do. It has three distinct variants:

- RGB565 – this is a good fit
- 8-bit indexed – this could be useful, as many UI images don't have that many colors. But the way format encodes the palette is inconvenient / inefficient
- RGB24/32 – not interesting for the same reason as QOI

Inconveniently, the bit stream for SLIC isn't documented yet. Fortunately, it is simple enough to figure out from the code.

I did a quick integration into Poco of the RGB565 mode. To make that memory efficient required rewriting much of the decoder in a way that is optimized for Poco's simultaneous decode / composite / display structure. It is working nicely. 

I'm thinking about taking a run at the 8-bit mode too. To be practical, that would first require lightly modifying the image format. I'm not sure whether the maintainer of SLIC would be open re-inteegrating something like that that though.

**Reply by wilberforce** [reply]
> To be practical, that would first require lightly modifying the image format. I'm not sure whether the maintainer of SLIC would be open re-inteegrating something like that that though.

I think it's worth asking!

**Reply by phoddie** [reply]
Yes, I agree. But I'll get it working first. I think it may sound like a more substantial change than it might actually be.

**Reply by phoddie** [reply]
I modified the SLIC reference implementation to handle palettes in a way that is faster and more memory friendly for RGB565 rendering. It can also be smaller, as it only stores the actual number of colors used rather than always reserving space for 256 colors. It isn't intended a complete solution, but just enough to experiment with. 

Using that I was able to encode images to use to implement 8-bit decoding. That works. I need to do tests to determine how well it performs, especially compared to RGB565 SLIC images.

**Reply by phoddie** [reply]
Did some more work on SLIC, this time to support alpha channel better. [Detailed notes](https://github.com/phoddie/SLIC/blob/dev/proposals.md) are available.

**Reply by wilberforce** [reply]
Have you sent the author your proposal? As his target is Arduino code he could be open to updating....

**Reply by phoddie** [reply]
Not yet. I'm planning to open an issue on that repository.

**Reply by phoddie** [reply]
Here's the [issue](https://github.com/bitbank2/SLIC/issues/4). The notes are updated as well.
